server/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java

```diff
@@ -554,6 +558,21 @@ public Projector visitUpdateProjection(final UpdateProjection projection, Contex
     private <A> ShardDMLExecutor<?, ?, A, ?> buildUpdateShardDMLExecutor(
         Context context, UpdateProjection projection,
         Collector<ShardResponse, A, Iterable<Row>> collector) {
+        // Get Stats to improve ram estimation for the update items
+        assert shardId != null : "ShardId must be provided for updates";
+        String indexName = shardId.getIndexName();
+        RelationName relationName = RelationName.fromIndexName(indexName);
+        final Stats stats = nodeCtx.tableStats().getStats(relationName);
+        final Collection<Reference> ramAccountedColumns;
+        // if stats are not available we fall back to estimate the size based on
+        // column types. Therefore we need to get the column information.
+        if (stats.isEmpty()) {
+            TableInfo tableInfo = nodeCtx.schemas().getTableInfo(relationName);
+            ramAccountedColumns = tableInfo.allColumns();
```

Comment:
allColumns doc:

```
/**
     * Top level and nested columns, including system columns. Excludes dropped columns
     **/
```



System columns can't be updated., we must exclude them.

I think they don't have stats but we will fall back to 
`stats == null `= > `UNKNOWN_DEFAULT_RAM_BYTES_USED` per sys column.

Reply:
> System columns can't be updated, we must exclude them.

In case of an update, the whole document is rewritten, including system columns, and send to the replica shard.

> I think they don't have stats but we will fall back to
> stats == null = > UNKNOWN_DEFAULT_RAM_BYTES_USED per sys column.

A fallback using an estimate based on the data type is already used, see https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/statistics/Stats.java#L154.

Reply:
To index an item, we also need the system columns in memory, such as `seq` or `id` etc. 
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java#L431C1-L437C38
This pr exists because `doc` is missing in the estimate. Some of the system columns are already estimated from `SHALLOW_SIZE` on the item, but some important onces such as `doc` are missing. I think it's better to overestimate than to underestimate. Finally, with https://github.com/crate/crate/issues/15059 this estimate will be rarely used anyway. So i would keep it the way it is.

Reply:
Right, thanks for the clarification!